### ISSUE TITLE
Razor regression in ASP .NET Core 9.0

### ISSUE THREAD

**Issue by AlexBAV** [dotnet/aspnetcore#58973]
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Describe the bug

The following razor snippet renders to nothing (is completely ignored??) in ASP.NET Core 9.0. Used to work correctly in previous version:

```html

<script type="importmap">
	 {
		 "imports": {
		 }
	}
</script>

```

Looks like a bug in server-side script tag helper.

### Expected Behavior

Should render "as is", without modifications

### .NET Version

9.0.100

**Comment by ahazelwood** [comment]
I would like to upgrade to 9.0, however this is hitting me on a asp.net core mvc project upgrading from net8.0.  Any suggestions for how to get it to render correctly would be greatly appreciated.  I concur that it should not override anything but render it as written.

**Comment by AlexBAV** [comment]
I currently use the following workaround for tag helpers that definitely should not be processed on a server side:

```html
<!script type="importmap">...</!script>
```

`<!tag>` syntax is an official way to turn off tag helper processing.

However, I'm still sure that tag helper should not remove a tag if it cannot process it on a server side.

**Comment by StevePy** [comment]
This one caught me as well, a bit sweaty when my importmap (part of _Layout.cshtml) wasn't rendered, considering the relatively good feeling I had combining razor views with ES6 modules. Fortunately the work-around with the `<!script type="importmap">` works a treat, but I can't see how that should be an expected breaking change since it's kind of a necessity for module scripts which are recognized. Fortunately a small work-around and pretty much everything else with the switch to .Net 9 has been relatively painless.

**Comment by seangwright** [comment]
I just spent a lot of time trying to figure out why my perfectly valid `<script type="importmap">` element wasn't working and my scripts weren't loading.

It turned out the import map wasn't showing up in my HTML response from ASPNET Core...

When I add the `!` to the `<script>` element it renders but now VS Code has lost any concept of Razor formatting this file with the C# dev kit. 🤦🤦

**Comment by seangwright** [comment]
Here's the cause of the issue
https://github.com/dotnet/aspnetcore/blob/v9.0.2/src/Mvc/Mvc.TagHelpers/src/ScriptTagHelper.cs#L251

```csharp
public override void Process(TagHelperContext context, TagHelperOutput output)
{
    ArgumentNullException.ThrowIfNull(context);
    ArgumentNullException.ThrowIfNull(output);

    if (string.Equals(Type, "importmap", StringComparison.OrdinalIgnoreCase))
    {
        // This is an importmap script, we'll write out the import map and
        // stop processing.
        var importMap = ImportMap ?? ViewContext.HttpContext.GetEndpoint()?.Metadata.GetMetadata<ImportMapDefinition>();
        if (importMap == null)
        {
            // No importmap found, nothing to do.
            output.SuppressOutput();
            return;
        }

        // ...
    }

    // ...
}
```

Because the `ScriptTagHelper` targets [every `<script>` with a `type` attribute](https://github.com/dotnet/aspnetcore/blob/v9.0.2/src/Mvc/Mvc.TagHelpers/src/ScriptTagHelper.cs#L33C43-L33C60) it will always execute for a `<script type="importmap">` element.

Additionally, if you aren't using the import map feature that's part of Razor components then ASP.NET Core assumes the `<script type="importmap">` is invalid and suppresses the output!

Hey, guess what ASP.NET Core... you don't get to decide if valid HTML is rendered or not 😠.

**Comment by marchy** [comment]
Nice chase-down @seangwright.

This is definitely a big with the .NET 9 framework. Team please resolve this. We are NOT using the new static asset pipeline and the behaviour of the rendering of our website should not change by upgrading – this is an unintentional regression.

**Comment by donnyv** [comment]
The worst part about this bug is that this post doesn't come up as a solution when you Google it. A stackoverflow post comes up about using the  new `asp-importmap="@data"`. As @seangwright correctly said. Asp.net should not be changing how html elements are rendered or should be rendered. Unless a "asp-*" attribute is added. Can't believe this got by testing. Spent way to long on this issue.

**Comment by seangwright** [comment]
@AlexBAV could you change the title of this issue to mention "importmap"? It might make it easier for others to find.

**Comment by AlexBAV** [comment]
> [@AlexBAV](https://github.com/AlexBAV) could you change the title of this issue to mention "importmap"? It might make it easier for others to find.

Is it better now?

**Comment by cjh826** [comment]
I'm also encountering this issue. In my Razor page, the content of `<script type="importmap">` gets replaced with either `{}` or automatically generated code  by the system like this in the browser:

```html
<script type="importmap">
{
  "imports": {
    "./js/site.js": "./js/site.xtxxf3hu2r.js",
    "./lib/bootstrap/dist/js/bootstrap.bundle.js": "./lib/bootstrap/dist/js/bootstrap.bundle.6cfz1n2cew.js",
    "./lib/bootstrap/dist/js/bootstrap.bundle.min.js": "./lib/bootstrap/dist/js/bootstrap.bundle.min.493y06b0oq.js",
    [...]
  }
}
</script>
```

 Whether it becomes `{}` or that long automatically generated list depends on these lines in program.cs:

```csharp
app.MapStaticAssets();
//app.UseStaticFiles();
app.MapRazorPages().WithStaticAssets();
//app.MapRazorPages();
```

Back in the ASP.NET Web Forms days, Microsoft tended to do too much behind the scenes, like a black box, making it hard for developers to understand what was actually happening and consequently difficult to implement desired functionality. While I certainly appreciate powerful framework capabilities, I much prefer having complete control as a developer rather than being left in the dark and unable to make changes. 

This current importmap issue, along with methods like `app.MapStaticAssets()` that don't accept any parameters (which would allow developers to specify behavior), is frustrating. I don't know what it's doing, and even if I did, I can't change it. Moreover, importmap is fundamentally a frontend technology - the backend (ASP.NET Core) shouldn't be doing anything with it unless explicitly instructed by the developer. I strongly dislike Microsoft's approach in this regard.

**Comment by cjh826** [comment]
I tried it, and this approach works:  
`@Html.Raw("<script type=\"importmap\">{\"imports\":{\"test\":\"test.js\"}}</script>")`"

**Comment by cjh826** [comment]
> I would like to upgrade to 9.0, however this is hitting me on a asp.net core mvc project upgrading from net8.0. Any suggestions for how to get it to render correctly would be greatly appreciated. I concur that it should not override anything but render it as written.



> Asp.net should not be changing how html elements are rendered or should be rendered. Unless a "asp-*" attribute is added.

Microsoft's approach caused them significant losses in the past, leaving their ecosystem far behind Java's. Yet they still cling to this bad habit of overstepping boundaries.

**Comment by MackinnonBuck** [comment]
@javiercn, maybe we should update the tag helper to not replace the content of an import map if it's explicitly provided?

**Comment by AlexBAV** [comment]
> [@javiercn](https://github.com/javiercn), maybe we should update the tag helper to not replace the content of an import map if it's explicitly provided?

Please also check if it fixes broken `<script>` tags with other types. When ASP.NET Core 9.0 broke `<script type=importmap>` script tags, it also broke `<script type="application/ld+json">` for us.

**Comment by danmoseley** [comment]
@MackinnonBuck I see you've assigned copilot, is this something we're pushing to get in RC2

**Comment by MackinnonBuck** [comment]
@danmoseley yeah, it would be nice to get this in. I made some adjustments to what copilot proposed, so I'll get some reviews and see if we can merge/backport it.

**Comment by kudorgyozo** [comment]
> I currently use the following workaround for tag helpers that definitely should not be processed on a server side:
> 
> <!script type="importmap">...</!script>
> `<!tag>` syntax is an official way to turn off tag helper processing.
> 
> However, I'm still sure that tag helper should not remove a tag if it cannot process it on a server side.

Thank you! You saved me so hard with this!
